ThirdPartyAdapters/unity/unity/src/main/java/com/google/ads/mediation/unity/UnityAdapter.java

```diff
@@ -190,60 +242,95 @@ public void requestInterstitialAd(Context context,
             }
             return;
         }
-        Activity activity = (Activity) context;
+        final Activity activity = (Activity) context;
         mActivityWeakReference = new WeakReference<>(activity);
 
-        UnitySingleton.getInstance().initializeUnityAds(activity, gameId);
-        UnitySingleton.getInstance().loadAd(mUnityAdapterDelegate);
-
+        UnityAds.addListener(mUnityAdapterDelegate);
+        UnitySingleton.getInstance().initializeUnityAds(activity, gameId,
+                new UnitySingleton.Listener() {
+                    @Override
+                    public void onInitializeSuccess() {
+                        MetaData metadata = new MetaData(activity);
+                        metadata.setCategory("mediation_adapter");
+                        metadata.set(uuid, "load-interstitial");
+                        metadata.set(uuid, mPlacementId);
+                        metadata.commit();
+
+                        UnitySingleton.getInstance().mPlacementsInUse.add(mPlacementId);
+                        UnityAds.load(mPlacementId);
+                    }
+
+                    @Override
+                    public void onInitializeError(String message) {
+                        mediationInterstitialListener.onAdFailedToLoad(UnityAdapter.this, AdRequest.ERROR_CODE_INTERNAL_ERROR);
+                    }
+                });
     }
 
     @Override
     public void showInterstitial() {
-        // Unity Ads does not have an ad opened callback. Sending Ad Opened event before showing the
-        // ad.
-        mMediationInterstitialListener.onAdOpened(UnityAdapter.this);
-
         if (mActivityWeakReference != null && mActivityWeakReference.get() != null) {
-            // Request UnitySingleton to show interstitial ads.
-            UnitySingleton.getInstance().showAd(mUnityAdapterDelegate, mActivityWeakReference.get());
+            MetaData metadata = new MetaData((mActivityWeakReference.get()));
+            metadata.setCategory("mediation_adapter");
+            metadata.set(uuid,"show-interstitial");
+            metadata.set(uuid, mPlacementId);
+            metadata.commit();
+
+            UnitySingleton.getInstance().mPlacementsInUse.remove(mPlacementId);
+            UnityAds.show(mActivityWeakReference.get(), mPlacementId);
         } else {
             Log.w(TAG, "Failed to show Unity Ads Interstitial.");
+            mMediationInterstitialListener.onAdOpened(UnityAdapter.this);
             mMediationInterstitialListener.onAdClosed(UnityAdapter.this);
         }
     }
+    //endregion
 
+    //region MediationAdapter implementation.
     @Override
     public void onDestroy() {
+        MetaData metadata = new MetaData((mActivityWeakReference.get()));
```

Comment:
should we remove placement id from  `UnitySingleton.getInstance().mPlacementsInUse`?